pkg/generate/cspc.go

```diff
@@ -0,0 +1,214 @@
+/*
+Copyright 2020-2021 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generate
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	cstorv1 "github.com/openebs/api/v2/pkg/apis/cstor/v1"
+	"github.com/openebs/api/v2/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/openebsctl/pkg/client"
+	"github.com/openebs/openebsctl/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// isPoolValid checks if a CStor pool is valid
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Pool calls the generate routine for different cas-types
+func Pool(nodes []string, devs int, raid string) error {
+	c := client.NewK8sClient()
+	if !isPoolTypeValid(strings.ToLower(raid)) {
+		// TODO: Use the well defined pool constant types from openebs/api when added there
+		return fmt.Errorf("invalid pool type %s", raid)
+	}
+	_, str, err := CSPC(c, nodes, devs, strings.ToLower(raid))
+	if err != nil {
+		return err
+	}
+	fmt.Println(str)
+	return nil
+}
+
+// CSPC takes eligible nodes, number of devices and poolType to create a pool cluster template
+func CSPC(c *client.K8sClient, nodes []string, devs int, poolType string) (*cstorv1.CStorPoolCluster, string, error) {
+	// 0. Figure out the OPENEBS_NAMESPACE for CStor
+	cstorNS, err := c.GetOpenEBSNamespace(util.CstorCasType)
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
```

Comment:
```suggestion
		return nil, "", fmt.Errorf("unable to determine the cStor namespace error: %v", err)
```
We are swallowing the original error in [c.GetOpenEBSNamespace](https://github.com/openebs/openebsctl/blob/9b279df44ee1c8c9d69950b08015690c8a0944b2/pkg/client/k8s.go#L188) good to return different error based on err variable... Something like below:


```go
func (k K8sClient) GetOpenEBSNamespace(casType string) (string, error) {
	pods, err := k.K8sCS.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{FieldSelector: "status.phase=Running", LabelSelector: fmt.Sprintf("openebs.io/component-name=%s", util.CasTypeAndComponentNameMap[strings.ToLower(casType)])})
	if err != nil {
		return "", errors.Wrapf(err, "unable to determine openebs namespace")
	}
        if len(pods.Items) == 0 {
                return "", errors.New("unable to find cStor related artifacts")
        }
	return pods.Items[0].Namespace, nil
}
```

Reply:
given the context is an empty one, i.e. it'll get stuck sort of forever and not get cancelled, what other `err` can come when `len(pod.Items)` is non zero?

Reply:
Ah, got it, even when the pods won't exits, the `err` will likely be nil, I'll make the necessary changes.